[scudo] Add EnableMultiRegions mode

ChiaHungDuan opened the pull request:
Instead of having single region for a size class, this mode increases
    the number of regions when one is exhausted. This reduces the
    fragmentation issue when one region is exhausted and gives the finer
    grunularity of choosing the size classes. However, it inevitably
    introduces some performance overhead because of the management of
    several regions for single size class. Currently, it's an experimental
    option and expected to see performance turbulance for a little bit.
    
    See more details and constraints in allocator_config.def.

ChiaHungDuan committed:
[scudo] Add EnableMultiRegions mode

Instead of having single region for a size class, this mode increases
the number of regions when one is exhausted. This reduces the
fragmentation issue when one region is exhausted and gives the finer
grunularity of choosing the size classes. However, it inevitably
introduces some performance overhead because of the management of
several regions for single size class. Currently, it's an experimental
option and expected to see performance turbulance for a little bit.

See more details and constraints in allocator_config.def.

ChiaHungDuan committed:
[scudo] Add support of assigning the alignment in ReservedMemory

ChiaHungDuan commented:
@cferris1000, there are two commits in this pull request. I'll split them up later. Now it's used for understanding the dependency

cferris1000 commented on compiler-rt/lib/scudo/standalone/allocator_config.def:
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
+
+// This allows each size class to have multiple regions instead of one. Note
+// that this is an experimental option so it has a few constraints while using.
+//   a. Pointer compaction is diabled. Which means `CompactPtrT` needs to be the
```
diabled -> disabled

to be the -> to be a

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/allocator_config.def:
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
+//      pointer integral type, i.e., uptr.
+//   b. `EnableRandomOffset` needs to be false. Pointer grouping requires
+//      the beginning of allocation address of a region to be aligned with
+//      `GroupSizeLog`. Without pointer compaction, it relies the region to be
```
relies -> requires

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/allocator_config.def:
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
+//   b. `EnableRandomOffset` needs to be false. Pointer grouping requires
+//      the beginning of allocation address of a region to be aligned with
+//      `GroupSizeLog`. Without pointer compaction, it relies the region to be
+//      allocated with proper alignment and the random offset will break the
```
and the random offset -> and a random offset

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/allocator_config.def:
```diff
@@ -84,6 +84,20 @@ PRIMARY_OPTIONAL(const uptr, CompactPtrScale, SCUDO_MIN_ALIGNMENT_LOG)
+//      allocated with proper alignment and the random offset will break the
+//      assumption.
+//   c. Condition variable is not supported under this mode. This is still under
+//      developing.
```
developing -> development

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/mem_map_linux.cpp:
```diff
@@ -127,11 +127,26 @@ void MemMapLinux::releaseAndZeroPagesToOSImpl(uptr From, uptr Size) {
+  if (!MemMap.map(Addr, MapSize, Name, Flags | MAP_NOACCESS))
     return false;
 
+  if (Alignment != getPageSizeCached()) {
```
Is it worth adding a comment indicating that this is not the normal path?

ChiaHungDuan replied:
Add `LIKELY`, let me know if you think a comment is better.

cferris1000 commented on compiler-rt/lib/scudo/standalone/mem_map_base.h:
```diff
@@ -93,9 +93,11 @@ template <class Derived, typename MemMapTy> class ReservedMemory {
   // Reserve a chunk of memory at a suggested address.
-  bool create(uptr Addr, uptr Size, const char *Name, uptr Flags = 0) {
+  bool create(uptr Addr, uptr Size, const char *Name, uptr Flags = 0,
+              uptr Alignment = getPageSizeCached()) {
```
Would it be better to change the alignment to be something like AlignmentPages and have the default be 1?

A lot of the places are checking that alignment is equal to getPageSizeCached() which is free on Android (most of the time), but is not as free on devices that don't have a hard-coded page size.

ChiaHungDuan replied:
This is a good idea. Done.

BTW, I haven't rebased so I haven't changed to use `getPageSizeLogCached` accordingly. Will do it later.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -56,6 +56,17 @@ template <typename Config> class SizeClassAllocator64 {
+  // Local cache stores the pointers in the type of compacted pointer and the
+  // compaction is done by calculating the offset to the base address of a
+  // region. Currently, we don't support decompacting through multiple regions
+  // because of the concern of performance and so we disable the pointer
```
because of ... -> due to performance concerns, we disable pointer compaction.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -56,6 +56,17 @@ template <typename Config> class SizeClassAllocator64 {
+  static const bool DisablePtrCompaction = Config::getEnableMultiRegions();
+  static_assert(!DisablePtrCompaction || sizeof(CompactPtrT) == sizeof(uptr),
+                "Pointer compaction is disabled, `CompactPtrT` needs to be the "
+                "same size of `uptr`");
```
same size of -> same size as

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -117,35 +128,45 @@ template <typename Config> class SizeClassAllocator64 {
+      // Block grouping requires the base address of a Region to be aligned
+      // with GrouopSize and pointer is compacted according to the offset to the
+      // base of a region so it always meets the requirement. As a result when
+      // the compaction is disabled, it relies the base address to be aligned.
```
it relies the base address -> it relies on the base address

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -155,82 +176,111 @@ template <typename Config> class SizeClassAllocator64 {
-      {
-        ScopedLock ML(Region->MMLock);
-        MemMapT MemMap = Region->MemMapInfo.MemMap;
+      auto RegionInfoIter = RegionInfoManager.getRegionInfoIter(I);
```
Would this be better as a for loop rather than a do while loop?

ChiaHungDuan replied:
There must be at least one RegionInfo instance (for multi-regions case, we allocate one in init()) so I take do-while loop when iterating RegionInfos. Let me know if you think it's still better for understanding (I'm fine with either way)

cferris1000 replied:
Seems reasonable to me.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -155,82 +176,111 @@ template <typename Config> class SizeClassAllocator64 {
-        for (const auto &It : BG.Batches)
-          TotalBlocks += It.getCount();
-      }
+      auto RegionInfoIter = RegionInfoManager.getRegionInfoIter(I);
```
Same as above, this seems like it would be a bit easier to understand as a for loop.

cferris1000 replied:
As above, your explanation seems reasonable.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -274,8 +347,24 @@ template <typename Config> class SizeClassAllocator64 {
+    };
+
+    // When multiple-regions is enabled, we need to sort the array to dispatch
+    // the blocks to different regions efficiently. Thus even we don't put
```
Thus even we -> If we

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -274,8 +347,24 @@ template <typename Config> class SizeClassAllocator64 {
+    // the blocks to different regions efficiently. Thus even we don't put
+    // BatchClass into groups, sorting is still necessary and it'll be handled
+    // later in the function.
+    // TODO: Reorder the use of variable
```
Is there some more to this TODO? I don't quite understand what this is referencing.

ChiaHungDuan replied:
Sorry, that's some private note for myself and which has been done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -408,15 +545,20 @@ template <typename Config> class SizeClassAllocator64 {
+    auto RegionInfoIter = RegionInfoManager.getRegionInfoIter(ClassId);
+
+    do {
+      // Note that the tryLock() may fail spuriously, given that it should
```
This comment is slightly hard to follow. Maybe something like:

Note that the tryLock() can fail under certain circumstances. Since this should be a rare occurrence, there is no need to do anything to force at least one page to be released.

ChiaHungDuan replied:
Change the description to what you suggested and with minor change in the end. Which is
```
... there is no need to do anything to force at least `releaseToOSMaybe()` is called().
```

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -425,36 +567,54 @@ template <typename Config> class SizeClassAllocator64 {
   static BlockInfo findNearestBlock(const char *RegionInfoData,
                                     uptr Ptr) NO_THREAD_SAFETY_ANALYSIS {
+    if (Config::getEnableMultiRegions()) {
+      Printf("MultiRegions hasn't supported finding nearest block yet.\n");
```
hasn't -> doesn't

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
+          return nullptr;
+        return &Array[Size++];
+      }
+      // The amount memory used by this allocator is about (NumEntries *
```
The amount memory -> The amount of memory

ChiaHungDuan replied:
Done

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
+        return &Array[Size++];
+      }
+      // The amount memory used by this allocator is about (NumEntries *
+      // RegionSize). For example, region with size 256 KB will have 2GB space
```
region with size 256 KB -> a 256 KB region

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
+                            RegionInfoLock[ClassId]);
+    }
+
+    // RegionInfos for the same size class will be stored in the order of base
```
store in the order of base  address -> ordered by base address

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
+      DCHECK_LT(ClassId, NumClasses);
+      DCHECK(Region->MemMapInfo.MemMap.isAllocated());
+
+      // The creation of new region requires holding the MMLock of current
```
of new region -> of a new region

of current -> of the current

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
+             sizeof(RegionInfoPointer) * NumClasses);
+    }
+
+    // Scudo requires the data member constant initializable. Array of raw
```
Array of raw -> An array of raw

ChiaHungDuan replied:
Done.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -555,18 +715,208 @@ template <typename Config> class SizeClassAllocator64 {
+    }
+
+    // Scudo requires the data member constant initializable. Array of raw
+    // pointers doesn't meet the condition. Therefore, wrap the pointer in the
```
meet the condition -> meet this condition

ChiaHungDuan replied:
Done.

ChiaHungDuan committed:
Fix the description in the comment and changed to alignment pages

ChiaHungDuan committed:
Change Alignment to AlignmentPages

ChiaHungDuan committed:
Address review comments

cferris1000 commented on compiler-rt/lib/scudo/standalone/mem_map_linux.cpp:
```diff
@@ -127,10 +127,29 @@ void MemMapLinux::releaseAndZeroPagesToOSImpl(uptr From, uptr Size) {
+  uptr MapSize = Size;
+
+  if (LIKELY(!NeedToAdjustAlignment)) {
+    if (!MemMap.map(Addr, MapSize, Name, Flags | MAP_NOACCESS))
```
Should this have MAP_ALLOWNOMEM in the flags? Same with theMap calls below.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -117,35 +127,46 @@ template <typename Config> class SizeClassAllocator64 {
     if (Config::getEnableContiguousRegions()) {
       ReservedMemoryT ReservedMemory = {};
+      // Block grouping requires the base address of a Region to be aligned
+      // with GrouopSize and pointer is compacted according to the offset to the
```
GrouopSize -> GroupSize

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -117,35 +127,46 @@ template <typename Config> class SizeClassAllocator64 {
+      // the compaction is disabled, it relies on the base address to be
+      // aligned.
+      const uptr Alignment =
+          DisablePtrCompaction ? (1UL << GroupSizeLog) : PageSize;
```
Is there a way to add a DCHECK that would fail if CompactPtrT is set but pointer compaction is disabled?

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -155,82 +176,111 @@ template <typename Config> class SizeClassAllocator64 {
-    if (conditionVariableEnabled()) {
+    RegionInfo *Region = RegionInfoManager.getCurRegionInfo(ClassId);
+
+    // TODO(chiahungduan): Support multiple-regions with condition variable.
```
Similar question, can we add  DCHECK that will fail if condition is enabled but isn't supported.

Better if it could be a static_assert to cause a compilation failure.

cferris1000 commented on compiler-rt/lib/scudo/standalone/primary64.h:
```diff
@@ -408,15 +544,20 @@ template <typename Config> class SizeClassAllocator64 {
+    do {
+      // Note that the tryLock() can fail under certain circumstances. Since
+      // this should be a rare occurrence, there is no need to do anything to
+      // force at least one `releaseToOSMaybe()` is called().
```
This is slightly awkward. Hope about something like:

force at least one call to `releaseToOSMaybe()`.

ChiaHungDuan committed:
[scudo] Make block storage in TransferBatch trailing objects

This allows us to change the number of blocks stored according to the
size of BatchClass.

Also change the name `TransferBatch` to `Batch` given that it's never
the unit of transferring blocks.

cferris1000 commented:
I've got a merged version of this that I'll be uploading sometime soon. So closing this one.
